torch/distributed/_tensor/placement_types.py

```diff
@@ -54,37 +54,102 @@ def _split_tensor(
-        idx_start_to_pad = tensor.size(self.dim) % num_chunks
-        if with_padding or contiguous:
+
+        # chunk tensor over dimension `dim` into n slices with padding if necessary
+        tensor_list = list(torch.chunk(tensor, num_chunks, dim=self.dim))
+        # compute the chunk size inline with ``torch.chunk``
+        full_chunk_size = (tensor.size(self.dim) + num_chunks - 1) // num_chunks
+
+        # Compute chunk size for each chunk for ``self.dim``
+        chunk_sizes = [
+            tensor_list[idx].size(self.dim) if idx < len(tensor_list) else 0
+            for idx in range(num_chunks)
+        ]
+        # Get idx start to pad
+        idx_start_to_pad = next(
+            (
+                idx
+                for idx, _ in enumerate(chunk_sizes)
+                if idx > 0 and chunk_sizes[idx] < chunk_sizes[idx - 1]
+            ),
+            0,
+        )
+        # Compute pad size on each chunk
+        pad_sizes = [
+            full_chunk_size - chunk_size if idx >= idx_start_to_pad else 0
+            for idx, chunk_size in enumerate(chunk_sizes)
+        ]
+
+        # Fill empty chunk with empty tensor
+        num_empty_tensors = num_chunks - len(tensor_list)
+        device = torch.device(
+            tensor.get_device() if torch.cuda.is_available() else "cpu"
+        )
+        empty_tensor = torch.tensor([], device=device)
+        for _ in range(num_empty_tensors):
+            tensor_list.append(empty_tensor)
+
+        if idx_start_to_pad != -1 and with_padding or contiguous:
             shard_list = []
-            for i, shard in enumerate(tensor_list):
-                if with_padding and idx_start_to_pad != 0 and i >= idx_start_to_pad:
-                    shard = self._pad_tensor(shard)
-                # input tensors are expected to be contiguous by the collective backend
+            for shard, pad_size in zip(tensor_list, pad_sizes):
+                # Fill the empty tensor with zeroes with padding.
+                if with_padding and pad_size == full_chunk_size:
+                    shard = self._pad_tensor(shard, pad_size, tensor_list[0])
+                elif with_padding and pad_size > 0:
+                    shard = self._pad_tensor(shard, pad_size)
                 shard = shard.contiguous() if contiguous else shard
                 shard_list.append(shard)
-            return shard_list, idx_start_to_pad
+            return shard_list, idx_start_to_pad, pad_sizes
         else:
-            return tensor_list, idx_start_to_pad
+            return tensor_list, idx_start_to_pad, pad_sizes
 
-    def _pad_tensor(self, tensor: torch.Tensor) -> torch.Tensor:
-        # pad tensor by 1 on the shard dim
-        pad = [0, 0] * (tensor.ndim - self.dim)
-        pad[-1] = 1
-        return torch.nn.functional.pad(tensor, pad)
+    def _pad_tensor(
+        self,
+        tensor: torch.Tensor,
+        pad_size: int,
+        reference_tensor: Optional[torch.Tensor] = None,
```

Comment:
I think we probably don't need this `reference_tensor` anymore if we creating the empty tensor with sth like `tensor.new_zeros((0, 3))`, we only need the pad_size and that would make the padding logic consistent too.